rspace/src/main/scala/coop/rchain/rspace/ReplayRSpace.scala

```diff
@@ -58,325 +55,345 @@ class ReplayRSpace[F[_], C, P, E, A, R, K](store: IStore[C, P, A, K], branch: Br
+              }
+      } yield r
 
-    @tailrec
     @SuppressWarnings(Array("org.wartremover.warts.Throw"))
     // TODO stop throwing exceptions
-    def getCommOrDataCandidates(comms: Seq[COMM]): Either[COMM, Seq[DataCandidate[C, R]]] =
-      comms match {
-        case Nil =>
-          val msg = "List comms must not be empty"
-          logger.error(msg)
-          throw new IllegalArgumentException(msg)
-        case commRef :: Nil =>
-          runMatcher(commRef) match {
-            case Some(x) => Right(x)
-            case None    => Left(commRef)
-          }
-        case commRef :: rem =>
-          runMatcher(commRef) match {
-            case Some(x) => Right(x)
-            case None    => getCommOrDataCandidates(rem)
-          }
-      }
-
-    logger.debug(s"""|consume: searching for data matching <patterns: $patterns>
-                     |at <channels: $channels>""".stripMargin.replace('\n', ' '))
-
-    val consumeRef = Consume.create(channels, patterns, continuation, persist, sequenceNumber)
-
-    replayData.get(consumeRef) match {
-      case None =>
-        span.mark("no-consume-ref-found")
-        Right(storeWaitingContinuation(consumeRef, None))
-      case Some(comms) =>
-        span.mark("ref-found")
-        val commOrDataCandidates: Either[COMM, Seq[DataCandidate[C, R]]] =
-          getCommOrDataCandidates(comms.iterator().asScala.toList)
-
-        commOrDataCandidates match {
-          case Left(commRef) =>
-            span.mark("no-data-candidates-found")
-            Right(storeWaitingContinuation(consumeRef, Some(commRef)))
-          case Right(dataCandidates) =>
-            span.mark("data-candidates-found")
-            Right(handleMatches(dataCandidates, consumeRef, comms))
+    def getCommOrDataCandidates(comms: Seq[COMM]): F[Either[COMM, Seq[DataCandidate[C, R]]]] = {
+      type COMMOrData = Either[COMM, Seq[DataCandidate[C, R]]]
+      def go(comms: Seq[COMM]) =
+        comms match {
+          case Nil =>
+            val msg = "List comms must not be empty"
+            //fixme replace with raiseError
+            logger.error(msg)
+            throw new IllegalArgumentException(msg)
+          case commRef :: Nil =>
+            runMatcher(commRef).map {
+              case Some(x) => x.asRight[COMM].asRight[Seq[COMM]]
+              case None    => commRef.asLeft[Seq[DataCandidate[C, R]]].asRight[Seq[COMM]]
+            }
+          case commRef :: rem =>
+            runMatcher(commRef).map {
+              case Some(x) => x.asRight[COMM].asRight[Seq[COMM]]
+              case None    => rem.asLeft[COMMOrData]
+            }
         }
+      comms.tailRecM(go)
     }
+
+    for {
+      _ <- logF.debug(s"""|consume: searching for data matching <patterns: $patterns>
+                     |at <channels: $channels>""".stripMargin.replace('\n', ' '))
+      consumeRef <- syncF.delay {
+                     Consume.create(channels, patterns, continuation, persist, sequenceNumber)
```

Comment:
I think we could've skipped delaying creation of those data structures - but I guess it's not immediately obvious there are no side-effects underneath, so better safe than sorry :+1:

Reply:
I chose to be consistent. If I hear a loud "hell no" then I'll revert.

Reply:
The readers might start to wander why the delay was done and start searching for side effects in `Consume.create`. This is misleading, i'd opt for removal